File: Swift/valid-palindrome-ii.swift
class Solution {
    func validPalindrome(_ s: String) -> Bool {
        let sChars = Array(s)
        var i = 0
        var j = s.count - 1

        while i < j {
            if sChars[i] != sChars[j] {
                return isPalindrome(sChars, i + 1, j) || isPalindrome(sChars, i, j - 1)
            } else {
                i += 1
                j -= 1
            }
        }

        return true
    }

    private func isPalindrome(_ s: [Character], _ i: Int, _ j: Int) -> Bool {
        var i = i
        var j = j

        while i < j {
            if s[i] != s[j] {
                return false
            } else {
                i += 1
                j -= 1
            }
        }

        return true
    }
}

File: Python/valid-palindrome-ii.py
class Solution:
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left = 0
        right = len(s)-1

        while left < right:
            if s[left] != s[right]:
                return self.isPalindrome(s, left, right-1) or self.isPalindrome(s, left+1, right)
            else:
                left += 1
                right -= 1

        return True

    def isPalindrome(self, s, left, right):
        while left < right:
            if s[left] != s[right]:
                return False
            else:
                left += 1
                right -= 1

        return True
